pkg/istio-agent/xds_proxy.go

```
@@ -319,6 +330,46 @@ func (p *XdsProxy) HandleUpstream(ctx context.Context, con *ProxyConnection, xds
 					TypeUrl:       v3.NameTableType,
 					ResponseNonce: resp.Nonce,
 				}
+			case v3.TrustBundleType:
+				if len(resp.Resources) == 0 {
+					return fmt.Errorf("empty response")
+				}
+				var tb v1.TrustBundleResponse
+				if err = ptypes.UnmarshalAny(resp.Resources[0], &tb); err != nil {
+					proxyLog.Errorf("failed to unmarshall trust bundles: %v", err)
+					return err
+				}
+				trustBundles := map[string][]byte{}
+				proxyLog.Debugf("received new trust bundles: %v", tb.TrustBundles)
+				expireTime := time.Date(3000, 1, 1, 1, 1, 1, 1, time.Now().Location())
+				for _, bundle := range tb.TrustBundles {
+					certBytes := []byte(bundle.RootCert)
+					certExpireTime, err := nodeagentutil.ParseCertAndGetExpiryTimestamp(certBytes)
+					if err != nil {
+						proxyLog.Errorf("failed to extract expiration time in the certificate loaded from file: %v", err)
+					} else if certExpireTime.Before(expireTime) {
+						expireTime = certExpireTime
+					}
+					trustBundles[bundle.TrustDomain] = certBytes
+				}
+
+				sc, ok := p.agent.WorkloadSecrets.(*cache.SecretCache)
+				if !ok || sc == nil {
+					proxyLog.Errorf("failed to access secret cache", err)
+					return nil
```

Should you return the error here?

---

I should. will fix